modules/tensor_mechanics/include/userobjects/SubblockIndexProvider.h

```diff
@@ -0,0 +1,40 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef SUBBLOCKINDEXPROVIDER_H
+#define SUBBLOCKINDEXPROVIDER_H
+
+#include "GeneralUserObject.h"
+
+class SubblockIndexProvider;
+
+template <>
+InputParameters validParams<SubblockIndexProvider>();
+
+/**
+ * Abstract base class for user objects that provide subblock scalar variable
+ * index
```

Comment:
I think I'd make this a little more descriptive.  Maybe something like:
Abstract base class for user objects that provide an index for a given element that is independent of the block id, so that behavior can be different on subsets of element blocks. This is used to apply independent generalized plane constraints to subsets of element blocks.